### ISSUE TITLE
Local Pickup is not charging taxes accurately (confirmed on bare bones WooCommerce installation)

### ISSUE THREAD

**Issue by saintcloud** [woocommerce/woocommerce#13144]
## EXPLANATION OF THE ISSUE

I downloaded WordPress 4.7.2 and installed WooCommerce 2.6.14.  I also installed the default theme "storefront" and then activated it.  I'm not using any other plugins.  I did this on Wamp, but I've also reproduced this error on Mamp, and on my cloud server on HG, and with different themes.  I've also reproduced this with several themes.  In short, I've isolated the problem to core WooCommerce (the notes below say to use twenty twelve as the theme, but it doesn't ship with WP anymore).  

When I have a billing address and shipping address both set to Utah, I choose "local pickup" as the shipping radio option, and I switch between billing and shipping ship-to addresses, there is no tax charged.  There should be a tax charged (it's the default functionality for local pickup).  However, if I choose "flat rate" then it shows up.  

Also, if I have a shipping address outside of the state of Utah, then there is a local pickup charged when switching between billing and shipping ship-to addresses.  

Again, this only shows up with choosing "local pickup" and when I have both the billing and shipping addresses set to Utah, which is the state of my local pickup.

This is similar to: https://github.com/woocommerce/woocommerce/issues/1491

## STEPS TO REPRODUCE THE ISSUE

**Setup:**
I added a single product.  I gave it a name, product description, regular price, sales price, **tax status = taxable, tax class = standard**, sku, weight, and dimensions.  Everything else was left empty or default.

I set up taxes:  
Tax options: https://screenshot.net/6lv00um
Standard Rates: https://screenshot.net/1or11u0

I set up Taxes:
Shipping Zones: https://screenshot.net/1o011ud
Shipping Options: https://screenshot.net/49dmmsr

I put in two address:
Billing: my address in Utah (but any Utah address will do)
Shipping: an address where I used to live in Florida (but any out of state address will do)

When having the "Ship to a different address?" checkbox checked, there is no tax and local pickup is not an option.  This is correct.  When I remove the "Ship to a different address?" checkbox, local pickup becomes an option (as it should) and having it selected produces a state tax (as it should): https://screenshot.net/w94oosw.  

Now, when I update the shipping address to be a different Utah address (any will do), the local pickup option will not show the state tax.  It doesn't matter if I choose the "ship to" address or the "billing address."  In both cases, state tax is a no show:   https://screenshot.net/r65oouz

If I switch between local pickup and flat rate, the tax does show up for flat rate for both addresses.

I should not have to have an out of state address to have state tax show up for local pickup.  If I choose to ship to my billing address with a Utah shipping address it won't work even if it's not used?  But if I chose to ship to my billing address with a Florida shipping address it will work, even if the Florida ship-to address isn't used?  That's a bug.

## SYSTEM STATUS REPORT

```
Grab the system status report from WooCommerce > System Status and paste it here.
```

`
### WordPress Environment ###

Home URL: http://localhost/woo
Site URL: http://localhost/woo
WC Version: 2.6.14
Log Directory Writable: ✔
WP Version: 4.7.2
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache/2.4.9 (Win64) PHP/5.5.12
PHP Version: ❌ 5.5.12 - We recommend a minimum PHP version of 5.6. See: How to update your PHP version
PHP Post Max Size: 3 MB
PHP Time Limit: 120
PHP Max Input Vars: 2500
cURL Version: 7.36.0
OpenSSL/1.0.1g

SUHOSIN Installed: –
MySQL Version: 5.6.17
Max Upload Size: 3 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 2.6.14
: 
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to \"Downloads\" and download the \"Binary / gzip\" file next to \"GeoLite Country\"

### Active Plugins (1) ###

WooCommerce: by WooThemes – 2.6.14

### Settings ###

Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2

### API ###

API Enabled: ✔

### WC Pages ###

Shop Base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My Account: #7 - /my-account/

### Taxonomies ###

Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

### Theme ###

Name: Storefront
Version: 2.1.7
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you're modifying WooCommerce on a parent theme you didn't build personally
then we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –
`

**Comment by mikejolley** [comment]
Local pickup changes which tax rates apply - see the documentation here https://docs.woocommerce.com/document/local-pickup/

You can make the location more specific using the code examples in that doc.

**Comment by saintcloud** [comment]
The billing address sometimes shows taxes with local pickup and sometimes it doesn't.  The criteria for whether or not it works?  The address for the unused shipping address.  Why does an unused shipping address determine whether or not the exact same billing address will work?  This is a bug.  It's obvious you didn't read the ticket.

**Comment by mikejolley** [comment]
Psychic? Local Pickup as I said charges taxes for your store location. Since your store location won't have a ZIP code unless you use the snippets in the doc I posted a link to, UTAH state tax will never show when local pickup is enabled. You said rates sometimes show, sometimes not, could just be the the checkout has not updated yet (e.g. because full address is not yet filled out).

Set the base location have a zip code starting with 84 and local puckup will charge the state tax. OR disable the local pickup tax rules again in the doc.

**Comment by saintcloud** [comment]
Utah state taxes are showing when local pickup is chosen (as I've outlined in the ticket).

Here is the **before** screenshot: https://screenshot.net/w94oosw.  It shows local pickup and taxes working for an address in Riverton, Utah (there is a zip code).  The ship-to address is unchecked.  What you don't see is that the ship-to address is for an address in Florida.  

Here is the **after** screenshot: https://screenshot.net/r65oouz.  It shows the local pickup and taxes **not working** for the exact same billing address in Riverton, Utah (again, with a zip code).  The ship-to address is unchecked.  What you don't see is that the ship-to address is for an address in Utah.

Why does having an unchecked ship-to address change whether or not the local pickup shows taxes for a billing address that didn't change?  The shipping address isn't even being used, and it determines whether or not the billing address will work? Again, the billing address never changed.  It is saved as an address.  And yet, sometimes it shows local pickup taxes and sometimes it doesn't, depending on the state of an unused shipping address.  Why?  This is a bug.

**Comment by mikejolley** [comment]
Have you used any of the snippets in that doc yet? Before I try to reproduce this.

**Comment by mikejolley** [comment]
For me, clean install, your setters, selecting local pickup gives no taxes which I expect because your tax rule is more specific than the store base location. Unless you're using a snippet there already.

**Comment by saintcloud** [comment]
I had the snippet in my functions.php file even before you mentioned it.  I had tested it with the snippet and it still wasn't working.  Just now I removed it and put it in again and now local pickup with taxes is working as expected.  I had a typo in my snippet. 

However, with the snippet not working, I was still getting local pickup rates when switching back and forth between billing and shipping addresses with an out of state (anything but Utah) shipping address.  Moving from the out of state shipping address to the local billing address gave me a local pickup state tax value.  I saw the spinning wheel calculating and putting in the state tax value when doing this.  Going back to the out of state shipping address removed the state tax  (as it should).  Then going back to the billing address put it back in again.  In short, it worked.

However, it did not do this if the shipping address was the same state as the billing address.  There is something weird going on there without the snippet that can't be explained, and I've been able to reproduce the weirdness in the default environment.  I had the snippet, so I had no reason to believe there was an issue with it because I was also getting local pickup tax rates in all cases except for when the un used shipping address was from the same state. 

Thanks for looking at this issue.

**Comment by mikejolley** [comment]
I updated my comment above. I'm not sure what code you had but just to clarify, BY DEFAULT when local pickup is selected, the user address is ignored for tax purposes and the shop address is used. This can only go as specific as country/state without code.

**Comment by saintcloud** [comment]
My original website had the snippet and had the tax/local pickup problem because the snippet was wrong.  I replicated the problem without the snippet on a fresh install, making me believe the problem had nothing to do with the snippet because it didn't work either way (or so I thought).  However, it is now working with the corrected snippet.  

I still believe there is a bug in there somewhere, even if the bug is negated by setting the base location with the snippet.

If a user doesn't have base location with the snippet, and switches from shipping (out-of-state) to billing (in-state), the local pickup tax doesn't have the base location but is still setting the local pickup location of the state or postal code of the billing address. In this case it is working.  But you say it shouldn't be doing this.  It shouldn't work, because how would it know that the billing address state is the local pickup state that has the state tax?

However, if the user is switching from shipping to billing and both address are in the same state, it doesn't use the state or zip code in the same way.  Why not?  Why is there different behavior?  If WooCommerce doesn't know what state/postal code is for local pickup without the snippet, then why does it create this in one instant but not another?

**Comment by mikejolley** [comment]
It's hard to follow what you're suggesting there, it's complex, but I do know the code explicitly converts to base location if local pickup is used. See https://github.com/woocommerce/woocommerce/blob/3dd6b483e5a37aed2cf10a2ccc2494bdec7563b1/includes/class-wc-customer.php#L185-L189

That uses base location only; not parts of shipping and billing. Only base. And base has no zip code unless you use code.

**Comment by saintcloud** [comment]
K, this code makes sense.  If it is "base," then use base country, state, postal code, etc.  

However, without having had "base" set with the snippet, I was still getting local pickup tax values.  How did it do this?  This part confused you too.  Were you not able to replicate this following my instructions?  Something in the code **is** getting billing address state or postal code (or something) when switching from out of state shipping to in state shipping and setting it to the base value.

I know this because it worked.  I got local pickup taxes **without** having the snippet,  and this could only have worked with it pulling from the billing address somehow.  And I could switch back and forth causing the script to run again, verifying it wasn't just sitting there from a previous address.  It was based on the billing address.  This should not be possible.  

Earlier I thought shipping wasn't working, but now it appears as if billing isn't working because it sets "base" properties somehow if the shipping address is out of state.  This doesn't make sense to me. Although, as I mentioned earlier, all this is negated with the snippet.

**Comment by mikejolley** [comment]
This only confused me from your posts above ^ On my own site I got the correct behaviour, thats why I assumed you had a snippet present.

**Comment by TazzySan** [comment]
I am having the same issue of not getting taxes on my local pick up. I put this snippet in the functions.php
add_filter( 'woocommerce_countries_base_postcode', create_function( '', 'return "49783";' ) );
add_filter( 'woocommerce_countries_base_city', create_function( '', 'return "SAULT STE MARIE";' ) );

 https://reflectionssoycandle.com/?page_id=102

**Comment by ChosenQuill** [comment]
Just a note. These code examples don't work on PHP V8 as create_function has been removed. 
You will need to create the function itself and add it into functions.php.

Example
```
function cs_postcode() {
	return '00000';
}
function cs_city() {
	return 'CITY HERE';
}

add_filter( 'woocommerce_countries_base_postcode', 'cs_postcode' );
add_filter( 'woocommerce_countries_base_city', 'cs_city' );
```

Maybe the docs need to be updated?
